sources/packages/api/src/services/application-exception/application-exception.service.ts

```diff
@@ -0,0 +1,176 @@
+import { Injectable } from "@nestjs/common";
+import { RecordDataModelService } from "../../database/data.model.service";
+import { Connection } from "typeorm";
+import {
+  Application,
+  ApplicationException,
+  ApplicationExceptionRequest,
+  ApplicationExceptionStatus,
+  Note,
+  NoteType,
+  Student,
+  User,
+} from "../../database/entities";
+import { CustomNamedError } from "../../utilities";
+import {
+  STUDENT_APPLICATION_EXCEPTION_INVALID_STATE,
+  STUDENT_APPLICATION_EXCEPTION_NOT_FOUND,
+} from "../../constants";
+
+/**
+ * Manages student applications exceptions detected upon full-time/part-time application
+ * submission, usually related to documents uploaded that must be reviewed.
+ */
+@Injectable()
+export class ApplicationExceptionService extends RecordDataModelService<ApplicationException> {
+  constructor(private readonly connection: Connection) {
+    super(connection.getRepository(ApplicationException));
+  }
+
+  /**
+   * Creates student application exceptions to be assessed by the Ministry.
+   * Exceptions are detected during full-time/part-time application submissions
+   * and are usually related to documents uploaded that must be reviewed.
+   * @param applicationId application that contains the exceptions.
+   * @param exceptionNames unique identifier names for the exceptions.
+   * @param auditUserId user that should be considered the one that is
+   * causing the changes.
+   * @returns created exception.
+   */
+  async createException(
+    applicationId: number,
+    exceptionNames: string[],
+    auditUserId: number,
+  ): Promise<ApplicationException> {
+    const creator = { id: auditUserId } as User;
+    const newException = new ApplicationException();
+    newException.application = { id: applicationId } as Application;
+    newException.creator = creator;
+    newException.exceptionStatus = ApplicationExceptionStatus.Pending;
+    newException.exceptionRequests = exceptionNames.map(
+      (exceptionName) =>
+        ({ exceptionName, creator } as ApplicationExceptionRequest),
+    );
+    return this.repo.save(newException);
+  }
+
+  /**
+   * Get a student application expectation detected after the student application was
+   * submitted, for instance, when there are documents to be reviewed.
+   * @param exceptionId exception to be retrieved.
+   * @returns student application expectation information.
+   */
+  async getExceptionById(exceptionId: number): Promise<ApplicationException> {
+    return this.repo
+      .createQueryBuilder("exception")
+      .select([
+        "exception.id",
+        "exception.exceptionStatus",
+        "exceptionNote.description",
+        "exceptionRequest.exceptionName",
+      ])
+      .innerJoin("exception.exceptionRequests", "exceptionRequest")
+      .innerJoin("exception.exceptionNote", "exceptionNote")
```

Comment:
Should we use innerJoin for exceptionNote? Unless the exception is approved/denied, it wouldn't have notes right?

Reply:
Fixed